### ISSUE TITLE
[Contributor Welcome] Implement C++ API version of torch.nn.MultiMarginLoss

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#27198]
## Context 

We would like to add the following two APIs to the C++ frontend:

*  `torch::nn::MultiMarginLoss` , which is the equivalent of Python API `torch.nn.MultiMarginLoss`.
* `torch::nn::functional::multi_margin_loss`, which is the equivalent of Python API `torch.nn.functional.multi_margin_loss`.

## Steps

* Add `torch::nn::MultiMarginLossOptions` to `torch/csrc/api/include/torch/nn/options/loss.h` (add this file if it doesn’t exist), which should include the following parameters (based on https://pytorch.org/docs/stable/nn.html#torch.nn.MultiMarginLoss)
    * `TORCH_ARG(int64_t, p) = 1`
    * `TORCH_ARG(double, margin) = 1.0`
    * `TORCH_ARG(Tensor, weight) = Tensor()`
    * `TORCH_ARG(Reduction::Reduction, reduction) = Reduction::Mean`
    * NOTE: please make sure to add the same comments for the parameters as in the Python version. For example, for parameter `p` we should say `Has a default value of 1. 1 and 2 are the only supported values`.
* Add `torch::nn::functional::multi_margin_loss(...)` in  `torch/csrc/api/include/torch/nn/functional/loss.h` (add this file if it doesn’t exist). The function should have the following signature:

```cpp
namespace torch {
namespace nn {
namespace functional {

inline Tensor multi_margin_loss(
    const Tensor& input,
    const Tensor& target,
    const MultiMarginLossOptions& options = {}) {
  ...
}

} // namespace functional
} // namespace nn
} // namespace torch
```

* Add `torch::nn::MultiMarginLoss` in `torch/csrc/api/include/torch/nn/modules/loss.h` (add this file if it doesn’t exist). The module’s constructor should take `torch::nn::MultiMarginLossOptions` as input (and store it in its internal options field), the module’s reset() method should register `weight` as a buffer (same logic as https://github.com/pytorch/pytorch/blob/5cac7387135ccf8db5b4b1ed9f52f80c924c2bc8/torch/nn/modules/loss.py#L20), and the module’s forward method should have the following signature:

```cpp
Tensor forward(
    const Tensor& input,
    const Tensor& target) {
  // Should call torch::nn::functional::multi_margin_loss with all inputs and the module's options
}
```

* Add test for `torch::nn::MultiMarginLoss` in test/cpp/api/modules.cpp. We can just check that the output value for some simple cases are as expected.

* Add test for `torch::nn::functional::multi_margin_loss(...)` in test/cpp/api/functional.cpp, using essentially the same tests as `torch::nn::MultiMarginLoss` .

## Helpful Resources

There are quite a few PRs for adding new functionals / new modules for the C++ API (the list of PRs is in https://github.com/pytorch/pytorch/issues/25883), which can serve as great references. Also please ping @yf225 on this issue if you encounter any problems.

## How do I claim this feature request?

Please comment in this issue if you are interested in working on it.

cc @yf225

**Comment by CarMiranda** [comment]
Hi @yf225, I would be glad to work on this issue!

**Comment by yf225** [comment]
@CarMiranda Thanks a lot for your help! Please let me know if you need anything from me :)

**Comment by heytanay** [comment]
I am interested on working this issue out, however, I am a beginner and need some directions as to understand the Source Code and start contributing. Any help will be highly appreciated! Thanks!

**Comment by yf225** [comment]
@heytanay Thanks a lot for your interest! @CarMiranda already contributed the `MultiMarginLoss` layer to libtorch (https://github.com/pytorch/pytorch/pull/27424), but please let me know if you are interested in working on any of the unclaimed layers listed in https://github.com/pytorch/pytorch/issues/25883 by leaving a comment there. :D The PRs listed there would be the best resources to get started, and https://github.com/pytorch/pytorch#from-source is the best instructions for building PyTorch. For running relevant libtorch tests, we can see the "How do I run tests?" section in https://github.com/pytorch/pytorch/issues/25883. Thanks!

**Comment by yf225** [comment]
Closed by https://github.com/pytorch/pytorch/pull/27424.
